### PR TITLE
otlLib documentation

### PR DESCRIPTION
**Pull request by simoncozens**

### CODE REVIEW

File: Lib/fontTools/otlLib/builder.py [hunk 1]
```diff
@@ -471,6 +839,15 @@ def add_subtable_break(self, location):
 
 
 class ClassPairPosSubtableBuilder(object):
+    """Builds class-based Pair Positioning (GPOS2 format 2) subtables.
+
+    Note that this does *not* build a GPOS2 ``otTables.Lookup`` directly,
+    but builds a list of ``otTables.PairPos`` subtables. It is used by the
+    :class:`PairPosBuilder` below.
+
+    Attributes:
```
Comment (anthrotype) [inline Lib/fontTools/otlLib/builder.py:848]: should we also document the other two init parameters `valueFormat1` and `valueFormat2` here?

File: Lib/fontTools/otlLib/builder.py [hunk 2]
```diff
@@ -668,6 +1204,20 @@ def buildLigatureSubstSubtable(mapping):
...
 
 def buildAnchor(x, y, point=None, deviceX=None, deviceY=None):
+    """Builds an Anchor table.
+
+    This determines the appropriate anchor format based on the passed parameters.
+
+    Args:
+        x (int): X coordinate.
+        y (int): Y coordinate.
+        point (int): Index of glyph contour point, if provided.
```
Comment (anthrotype) [inline Lib/fontTools/otlLib/builder.py:1214]: I don't know if we want to use the typing annotation syntax in the docstrings or not (ideally yes, but perhaps can be done later?), e.g.

```suggestion
        point (Optional[int]): Index of glyph contour point, if provided.
```

If we do, we'll have to change this and many other similar places.

File: Lib/fontTools/otlLib/builder.py [hunk 3]
```diff
@@ -695,14 +1270,29 @@ def buildBaseArray(bases, numMarkClasses, glyphMap):
 
 
 def buildBaseRecord(anchors):
-    """[ot.Anchor, ot.Anchor, ...] --> ot.BaseRecord"""
+    # [otTables.Anchor, otTables.Anchor, ...] --> otTables.BaseRecord
+    # (Not docstringed because it's only needed when building BaseArray,
+    # at which point you should be using the above.)
```
Comment (anthrotype) [inline Lib/fontTools/otlLib/builder.py:1275]: this comment is not necessary, we just stick to the implicit rule that if something hasn't a `"""` docstring it isn't in the publicly documented API

File: Lib/fontTools/otlLib/builder.py [hunk 4]
```diff
@@ -765,15 +1423,41 @@ def buildLigatureArray(ligs, numMarkClasses, glyphMap):
 
 
 def buildLigatureAttach(components):
-    """[[Anchor, Anchor], [Anchor, Anchor, Anchor]] --> LigatureAttach"""
+    # [[Anchor, Anchor], [Anchor, Anchor, Anchor]] --> LigatureAttach
+    # (Not docstringed because it's only needed when building LigatureArray,
+    # at which point you should be using the above.)
```
Comment (anthrotype) [inline Lib/fontTools/otlLib/builder.py:1428]: unnecessary comment, like above

File: Lib/fontTools/otlLib/builder.py [hunk 5]
```diff
@@ -863,14 +1643,16 @@ def buildMarkRecord(classID, anchor):
 
 
 def buildMark2Record(anchors):
-    """[ot.Anchor, ot.Anchor, ...] --> ot.Mark2Record"""
+    # [otTables.Anchor, otTables.Anchor, ...] --> otTables.Mark2Record
+    # (Not docstringed because it's only needed when building MarkMarkPos,
+    # at which point you should be using otl.lookupBuilders.MarkMarkPosBuilder.)
```
Comment (anthrotype) [inline Lib/fontTools/otlLib/builder.py:1648]: pls remove comment

File: Lib/fontTools/otlLib/builder.py [hunk 6]
```diff
@@ -854,6 +1633,7 @@ def buildMarkLigPosSubtable(marks, ligs, glyphMap):
 
 
 def buildMarkRecord(classID, anchor):
+    # Only used by buildMarkArray
```
Comment (anthrotype) [inline Lib/fontTools/otlLib/builder.py:1636]: is this comment necessary?

File: Lib/fontTools/otlLib/builder.py [hunk 7]
```diff
@@ -129,13 +184,39 @@ def build_subst_subtables(self, mapping, klass):
         return subtables
 
     def add_subtable_break(self, location):
+        """Add an explicit subtable break.
+
+        Args:
+            location: A string or tuple representing the location in the
```
[the remark this answers is not in the document]
Comment (simoncozens) [reply Lib/fontTools/otlLib/builder.py:190]: I don't like calling something  "optional" if you have to provide it. :-) I'll reword this.

File: Lib/fontTools/otlLib/builder.py (hunk 1 above)
Comment (simoncozens) [reply Lib/fontTools/otlLib/builder.py:848]: I can't see that they're actually used. I'm going to remove them instead...

Comment (anthrotype) [reply Lib/fontTools/otlLib/builder.py:848]: Interesting. I wonder why they were there in the first place..

**Comment by simoncozens** [comment]
We will rethink the typing information on the docs another time.

Are we happy with the change to ClassPairPosSubtableBuilder? If so I'll merge this.

**Comment by anthrotype** [comment]
> Are we happy with the change to ClassPairPosSubtableBuilder? If so I'll merge this.

it's a pretty niche class, if those attributes weren't really used internally so they can safely be dropped indeed.

**Pull request merged**